File: otimization ii/lasso.jl
# linear_regression_mse.jl

using LinearAlgebra, Statistics, GLM, DataFrames

function linear_regression_mse()
    # Generate data
    x = randn(100, 2)
    y = x * [2, 1] + randn(100) * 0.05
    
    # MSE function
    function mse(b1, b2)
        y_pred = x * [b1, b2]
        return mean((y .- y_pred) .^ 2)
    end
    
    # Coefficient sequences
    coef_seq = range(-1, 5, length=200)
    
    # Calculate MSE surface
    m = zeros(length(coef_seq), length(coef_seq))
    for i in 1:length(coef_seq)
        for j in 1:length(coef_seq)
            m[i,j] = mse(coef_seq[i], coef_seq[j])
        end
    end
    
    # L1 regularization levels
    l1(b1, b2) = abs(b1) + abs(b2)
    
    l1_levels = zeros(length(coef_seq), length(coef_seq))
    for i in 1:length(coef_seq)
        for j in 1:length(coef_seq)
            l1_levels[i,j] = l1(coef_seq[i], coef_seq[j])
        end
    end
    
    # OLS coefficients using GLM
    df = DataFrame(x1=x[:,1], x2=x[:,2], y=y)
    model = lm(@formula(y ~ 0 + x1 + x2), df)
    ols_coefs = coef(model)
    
    return (mse_surface=m, l1_levels=l1_levels, 
            coef_seq=coef_seq, ols_coefs=ols_coefs,
            true_coefs=[2, 1])
end

# Run the analysis
results = linear_regression_mse()
println("OLS coefficients: ", results.ols_coefs)
println("True coefficients: ", results.true_coefs)

File: otimization ii/lasso.py
# linear_regression_mse.py

import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

def linear_regression_mse():
    # Set random seed for reproducibility
    np.random.seed(42)
    
    # Generate data
    x = np.random.randn(100, 2)
    y = x @ np.array([2, 1]) + np.random.randn(100) * 0.05
    
    # MSE function
    def mse(b1, b2):
        y_pred = x @ np.array([b1, b2])
        return np.mean((y - y_pred) ** 2)
    
    # Coefficient sequences
    coef_seq = np.linspace(-1, 5, 200)
    
    # Create meshgrid for 2D evaluation
    B1, B2 = np.meshgrid(coef_seq, coef_seq)
    
    # Calculate MSE surface
    m = np.zeros_like(B1)
    for i in range(B1.shape[0]):
        for j in range(B1.shape[1]):
            m[i,j] = mse(B1[i,j], B2[i,j])
    
    # L1 regularization levels
    def l1(b1, b2):
        return np.abs(b1) + np.abs(b2)
    
    l1_levels = l1(B1, B2)
    
    # OLS coefficients
    model = LinearRegression(fit_intercept=False)
    model.fit(x, y)
    ols_coefs = model.coef_
    
    return {
        'mse_surface': m,
        'l1_levels': l1_levels,
        'B1': B1,
        'B2': B2,
        'ols_coefs': ols_coefs,
        'true_coefs': np.array([2, 1])
    }

if __name__ == "__main__":
    results = linear_regression_mse()
    print("OLS coefficients:", results['ols_coefs'])
    print("True coefficients:", results['true_coefs'])